r1cs/gadgets/std/src/bits/macros.rs

```diff
@@ -0,0 +1,2532 @@
+                        let mut multi_eq = MultiEq::new(&mut cs);
+                        $type_name::addmany(multi_eq.ns(|| "add operands"), &operands).unwrap()
+                    };
+
+                    test_uint_gadget_value(result_value, &result_var, "result correctness");
+                    assert!(cs.is_satisfied());
+
+                    // negative test
+                    let bit_gadget_path = "add operands/alloc result bit 0/boolean";
+                    if cs.get(bit_gadget_path).is_zero() {
+                        cs.set(bit_gadget_path, Fr::one());
+                    } else {
+                        cs.set(bit_gadget_path, Fr::zero());
+                    }
+                    assert!(!cs.is_satisfied());
+
+                    // test with all constants
+                    let mut cs = ConstraintSystem::<Fr>::new(SynthesisMode::Debug);
+                    let operands = operand_values.iter().enumerate().map(|(i, val)| {
+                        alloc_fn(&mut cs, format!("alloc constant operand {}", i).as_str(), &VARIABLE_TYPES[0], *val)
+                    }).collect::<Vec<_>>();
+                    let num_constraints = cs.num_constraints();
+
+                    let result_var = {
+                        // add a scope for multi_eq CS as the constraints are enforced when the variable is dropped
+                        let mut multi_eq = MultiEq::new(&mut cs);
+                        $type_name::addmany(multi_eq.ns(|| "add constant operands"), &operands).unwrap()
+                    };
+
+                    test_uint_gadget_value(result_value, &result_var, "sum of constants result correctness");
+                    assert!(cs.is_satisfied());
+                    assert_eq!(cs.num_constraints(), num_constraints);
+                }
+
+                #[test]
+                #[allow(unconditional_panic)] // otherwise test will not compile for uint128, as field is too small
+                fn test_mulmany() {
+                    const MAX_NUM_OPERANDS: usize = (<Fr as PrimeField>::Params::CAPACITY) as usize/$bit_size ;
+                    const NUM_OPERANDS: usize = MAX_NUM_OPERANDS*2+5;
+                    // we want to test a case when the operands must be split in multiple chunks
+                    assert!(NUM_OPERANDS > MAX_NUM_OPERANDS);
+
+
+                    let rng = &mut thread_rng();
+                    let mut cs = ConstraintSystem::<Fr>::new(SynthesisMode::Debug);
+
+                    let operand_values = (0..NUM_OPERANDS).map(|_| rng.gen()).collect::<Vec<$native_type>>();
+
+                    let operands = operand_values.iter().enumerate().map(|(i, val)| {
+                        alloc_fn(&mut cs, format!("alloc operand {}", i).as_str(), &VARIABLE_TYPES[i % 3], *val)
+                    }).collect::<Vec<_>>();
+
+                    let result_value: $native_type = operand_values.iter().map(|el| *el).reduce(|a,b| a.overflowing_mul(b).0).unwrap();
```

Comment:
You can replace `map(|el| *el)` with `copied()`.
This is done in several places.
Try to use `cargo clippy`, it will highlight such stuff

Reply:
Modified